### DISCUSSION TITLE
Question about providing liquidity and ticks

### DISCUSSION THREAD

**Discussion by ryang-max** [Jeiwan/uniswapv3-book#10] [Milestone 2. Second Swap] [answered]
When providing liquidity in a price range[a, b], the code only update the ticks of the lower bound and the upper bound, but not all ticks in this range, so do I really provide liquidity in the whole price range, or I just provide liquidity in the two ticks? It seems that when the current tick is between the lower bound and upper bound of the price range I provided, the swap does not use my liquidity at all.

Besides, since the liquidity is added to both two ticks, so does it mean I provide double liquidity? If a swap starts at a current price lower than the lower bound of my price range, and the target price higher than the upper bound of my price range, is my liquidity used twice?

**Comment by Jeiwan** [comment, marked as answer]
We don't need to store liquidity at each tick–that would've been very gas inefficient to store and update each of them. Instead, we only need to know the ranges that have liquidity and the current price:
1. if the current price is inside a range (or multiple ranges), there's liquidity for trades;
1. when all liquidity ranges are above or below the current price, there's no liquidity for trades.

You can see this idea in the code:
https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L226-L256
1. when liquidity is added [above](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L226-L231) the current price, the `liquidity` state variable of the pool (the $L$) is not updated;
1. the same is true when liquidity is added [below](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L250-L254) the current price;
1. however, when liquidity is added [around](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L232-L248) the current price, the $L$ is increased.

The `liquidity` state variable tracks the amount of liquidity **currently** available, i.e. the sum of liquidity of all the price ranges that include the current price. It [gets](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L478-L481) [updated](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L518) when the price is leaving and entering price ranges ([crossing a tick](https://github.com/Jeiwan/uniswapv3-code/blob/main/src/UniswapV3Pool.sol#L462)) during swapping (swapping changes price). It's for such cases that we store liquidity at the boundary ticks: when a boundary tick is crossed, we know how much liquidity to remove or add to the `liquidity` state variable. When the lower bound is crossed from left to right, liquidity of the tick (in fact, of the liquidity position it belongs to) is added; when the upper bound is crossed from left to right, liquidity of the tick (again, liquidity of the position) is removed. It's the other way around when price is falling.

These concept are explained and implemented in [Milestone 3](https://uniswapv3book.com/docs/milestone_3/introduction/).
